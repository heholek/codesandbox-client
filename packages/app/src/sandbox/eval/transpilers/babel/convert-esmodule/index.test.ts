import { convertEsModule } from '.';

describe('convert-esmodule', () => {
  it('can convert reexports', () => {
    const code = `
      export * from './test/store.js';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert normal exports', () => {
    const code = `
      export { test, test2 } from './test/store.js';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert function exports', () => {
    const code = `
      export function test() {}
      export const test2 = () => {}
      export class Test {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert default exports', () => {
    const code = `
      export default function test() {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert class default exports', () => {
    const code = `
      export default class A {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert weird default exports', () => {
    const code = `
      export default a = 'b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert default imports', () => {
    const code = `
      import a from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert named imports', () => {
    const code = `
      import {a, b, c} from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle as imports', () => {
    const code = `
      import { a as b } from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('ignores comments', () => {
    const code = `
      // import { a as b } from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle inline comments', () => {
    const code = `
      import { a as b, /* HELLO WORLD */ c } from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle class properties', () => {
    const code = `
      import T from './test';
      class T2 {
        a = () => {
          return 'test'
        }
        b = ''
        c = {}
        static d = ''
      }
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle async code', () => {
    const code = `
      import T from './test';
      async () => {
        const test = await test2()
      }
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles export mutations', () => {
    const code = `
      export default function test() {}

      test = 5;
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles export mutations with no named function', () => {
    const code = `
      export default function() {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles default as exports', () => {
    const code = `
    export { default as Field } from './Field';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('parses and writes chars with linebreaks', () => {
    const code =
      "var WS_CHARS = 'u2000-\\u200a\\u2028\\u2029\\u202f\\u205f\\u3000\\ufeff'";
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('has good perf', () => {
    /* eslint-disable */
    const code = require('./big-file');

    const t = Date.now();

    for (let i = 0; i < 1; i++) {
      convertEsModule(code);
    }
    console.log(Date.now() - t);
    /* eslint-enable */
  });
});
