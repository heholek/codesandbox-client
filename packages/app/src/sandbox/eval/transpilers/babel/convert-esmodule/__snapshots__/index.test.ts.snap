// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-esmodule can convert class default exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = A;
class A {}
"
`;

exports[`convert-esmodule can convert default exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = test;
function test() {}
"
`;

exports[`convert-esmodule can convert default imports 1`] = `
"var $csb__b = require(\\"./b\\");
var a = $_csb__interopRequireDefault($csb__b).default;
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can convert function exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function test() {}
exports.test = test;
const test2 = () => {};
exports.test2 = test2;
class Test {}
exports.Test = Test;
"
`;

exports[`convert-esmodule can convert named imports 1`] = `
"var $csb__b = require(\\"./b\\");
var a = $csb__b.a;
var b = $csb__b.b;
var c = $csb__b.c;
"
`;

exports[`convert-esmodule can convert normal exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__store = require(\\"./test/store.js\\");
exports.test = $csb__store.test;
exports.test2 = $csb__store.test2;
"
`;

exports[`convert-esmodule can convert reexports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__store = require(\\"./test/store.js\\");
Object.keys($csb__store).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  exports[key] = $csb__store[key];
});
"
`;

exports[`convert-esmodule can convert weird default exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__default = a = \\"b\\";
exports.default = $csb__default;
"
`;

exports[`convert-esmodule can handle as imports 1`] = `
"var $csb__b = require(\\"./b\\");
var b = $csb__b.a;
"
`;

exports[`convert-esmodule can handle async code 1`] = `
"var $csb__test = require(\\"./test\\");
var T = $_csb__interopRequireDefault($csb__test).default;
(async () => {
  const test = await test2();
});
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can handle class properties 1`] = `
"var $csb__test = require(\\"./test\\");
var T = $_csb__interopRequireDefault($csb__test).default;
class T2 {
  a = () => {
    return \\"test\\";
  }
  b = \\"\\"
  c = {}
  static d = \\"\\"
}
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can handle inline comments 1`] = `
"var $csb__b = require(\\"./b\\");
var b = $csb__b.a;
var c = $csb__b.c;
"
`;

exports[`convert-esmodule handles default as exports 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__Field = require(\\"./Field\\");
exports.Field = $csb__Field.default;
"
`;

exports[`convert-esmodule handles export mutations 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = test;
function test() {}
exports.default = test = 5;;
"
`;

exports[`convert-esmodule handles export mutations with no named function 1`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = $csb__default;
function $csb__default() {}
"
`;

exports[`convert-esmodule ignores comments 1`] = `""`;

exports[`convert-esmodule parses and writes chars with linebreaks 1`] = `
"var WS_CHARS = \\"u2000-     　﻿\\";
"
`;
